content/chapters/compute/lab/README.md

```diff
@@ -0,0 +1,685 @@
+Let's take a look at them.
+The numbers below will differ from those on your system:
+
+```
+student@os:~/.../support/sleepy:$ cat syscalls.2523393  # syscalls from parent process
+execve("sleepy_creator", ["sleepy_creator"], 0x7ffd2c157758 /* 39 vars */) = 0
+--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=2523394, si_uid=1052093, si_status=0, si_utime=0, si_stime=0} ---
++++ exited with 0 +++
+
+student@os:~/.../support/sleepy:$ cat syscalls.2523394  # syscalls from child process
+execve("/bin/sh", ["sh", "-c", "sleep 10"], 0x7ffd36253be8 /* 39 vars */) = 0
+execve("/usr/bin/sleep", ["sleep", "10"], 0x560f41659d40 /* 38 vars */) = 0
++++ exited with 0 +++
+```
+
+[Quiz](./quiz/who-calls-execve-parent.md)
+
+Now notice that the child process doesn't simply call `execve("/usr/bin/sleep" ...)`.
+It first changes its virtual address space (VAS) to that of a `bash` process (`execve("/bin/sh" ...)`) and then that `bash` process switches its VAS to `sleep`.
+Therefore, calling `system(<some_command>)` is equivalent to running `<some_command>` in the command line.
+
+With this knowledge in mind, let's implement our own mini-shell.
+Start from the skeleton code in `support/mini-shell/mini_shell.c`.
+We're already running our Bash interpreter from the command line, so there's no need to `exec` another Bash from it.
+Simply `exec` the command.
+
+[Quiz](./quiz/mini-shell-stops-after-command.md)
+
+So we need a way to "save" the `mini_shell` process before `exec()`-ing our command.
+Find a way to do this.
+
+> **Hint**:  You can see what `sleepy` does and draw inspiration from there.
+> Use `strace` to also list the calls to `clone()` perfromed by `sleepy` or its children.
+> [Remember](#threads-and-processes-clone) what `clone()` is used for and use its parameters to deduce which of the two scenarios happens to `sleepy`. 
+
+**Moral of the story**: We can add another step to the moral of [our previous story](#practice-fork).
+When spawning a new command, the call order is:
+- parent: `fork()`, `exec()`, `wait()`
+- child: `exit()`
+
+#### Command Executor in Another language
+
+Now implement the same functionality (a Bash command executor) in any other language, other than C/C++.
+Use whatever API is provided by your language of choice for creating and waiting for processes.
+
+### The GIL
+
+Throughout this lab you might have noticed that there were no thread exercises _in Python_.
+If you did, you probably wondered why.
+It's not because Python does not support threads, because it does, but because of a mechanism called the **Global Interpreter Lock**, or GIL.
+As its name suggests, this is a lock implemented inside most commonly used Python interpreter (CPython), which only allows **one** thread to run at a given time.
+As a consequence, multithreaded programs written in Python run **concurrently**, not in parallel.
+For this reason, you will see no speedup even when you run an embarrassingly parallel code in parallel.
```

Comment:
It should be mentioned that python threads are still useful when you have multiple io-bound tasks. Otherwise one would believe that they are completely useless.